File: app/src/main/java/io/github/zwieback/familyfinance/core/lifecycle/creator/EntityViewCreator.java
package io.github.zwieback.familyfinance.core.lifecycle.creator;

import androidx.annotation.NonNull;

import java.sql.Connection;
import java.sql.Statement;
import java.util.concurrent.Callable;

import io.reactivex.Single;

public abstract class EntityViewCreator implements Callable<Single<Boolean>> {

    private final Connection connection;

    protected EntityViewCreator(Connection connection) {
        this.connection = connection;
    }

    @NonNull
    protected abstract String getViewName();

    @NonNull
    protected abstract String getViewBody();

    /**
     * @return {@code false} because "CREATE VIEW" is DDL operation
     */
    @Override
    public Single<Boolean> call() {
        return Single.fromCallable(() -> {
            try (Statement statement = connection.createStatement()) {
                return statement.execute("CREATE VIEW IF NOT EXISTS " + getViewName() + " AS " +
                        getViewBody());
            }
        });
    }
}

File: app/src/main/java/io/github/zwieback/familyfinance/core/lifecycle/creator/EntityViewCreator.kt
package io.github.zwieback.familyfinance.core.lifecycle.creator

import io.reactivex.Single
import java.sql.Connection
import java.util.concurrent.Callable

abstract class EntityViewCreator protected constructor(
    private val connection: Connection
) : Callable<Single<Boolean>> {

    protected abstract val viewName: String

    protected abstract val viewBody: String

    /**
     * @return `false` because "CREATE VIEW" is DDL operation
     */
    override fun call(): Single<Boolean> {
        return Single.fromCallable {
            connection.createStatement().use { statement ->
                statement.execute("CREATE VIEW IF NOT EXISTS $viewName AS $viewBody")
            }
        }
    }
}
